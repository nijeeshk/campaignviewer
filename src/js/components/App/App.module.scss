.app {
  padding: 1rem;

  .container {
    width: 100%;
    margin-right: auto;
    margin-left: auto;

    @media (min-width: 576px) {
      max-width: 540px;
    }

    @media (min-width: 768px) {
      max-width: 720px;
    }

    @media (min-width: 992px) {
      max-width: 960px;
    }

    @media (min-width: 1200px) {
      max-width: 1140px;
    }
  }

  .panelHeader {
    display: flex;
    flex-direction: column;
    
    .left,
    .right {
      flex: 1;
      width: 100%;
      margin-bottom: 1rem;
    }

    @media (min-width: 768px) {
      flex-direction: row;
      justify-content: space-between;

      .left,
      .right {
        max-width: 280px;
        margin-bottom: 0;
      }
    }
  }
}
