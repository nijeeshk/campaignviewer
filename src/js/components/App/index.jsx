import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import uuidV4 from 'uuid/v4';
import { DateRangePicker } from 'react-dates';

import {
  Panel,
  PanelHeader,
  PanelBody,
} from '../../widgets/Panel';
import Input from '../../widgets/Input';
import Toast from '../../widgets/Toast';
import Table from '../../widgets/Table';
import processCampaigns from '../../utils/processCampaigns';
import 'react-dates/lib/css/_datepicker.css';
import 'react-dates/initialize';
import styles from './App.module.scss';

class App extends Component {
  static defaultProps = {
    campaigns: [],
    startDate: null,
    endDate: null,
    focusedInput: null,
    pageWidth: 0,
    toast: {
      id: '',
      text: '',
    },
  };

  static propTypes = {
    campaigns: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        startDate: PropTypes.string.isRequired,
        endDate: PropTypes.string.isRequired,
        Budget: PropTypes.number.isRequired,
      }),
    ).isRequired,
    startDate: PropTypes.object,
    endDate: PropTypes.object,
    focusedInput: PropTypes.string,
    pageWidth: PropTypes.number,
    toast: PropTypes.shape({
      id: PropTypes.string,
      text: PropTypes.string,
    }),
  };

  state = {
    campaigns: this.props.campaigns,
    startDate: this.props.startDate,
    endDate: this.props.endDate,
    focusedInput: this.props.focusedInput,
    pageWidth: this.props.pageWidth,
    toast: this.props.toast,
  };

  resizeHandle = null;

  componentDidMount() {
    window.AddCampaigns = this.addCampaigns;
    this.handleResize();
    window.addEventListener('resize', this.handleResize);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
  }

  handleResize = () => {
    if (this.resizeHandle) {
      clearTimeout(this.resizeHandle);
    }
    this.resizeHandle = setTimeout(() => {
      this.setState({
        pageWidth: document.documentElement.offsetWidth,
      });
    }, 500);
  }

  addCampaigns = (campaigns) => {
    const { list, failed } = processCampaigns(campaigns);
    const toastText = `Add Campaigns: ${list.length} success. ${failed.length} failed`;
    this.setState({
      campaigns: [
        ...this.state.campaigns,
        ...list,
      ],
      toast: {
        id: uuidV4(),
        text: toastText,
      },
    });
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.setState({
      startDate,
      endDate,
    });
  }

  onFocusChange = (focusedInput) => {
    this.setState({ focusedInput });
  }

  isOutsideRange = (date) => {
    const { startDate, endDate } = this.state;
    if (moment(startDate).isValid()
      && moment(endDate).isValid()) {
      return false;
    }
    if (moment(startDate).isValid()
      && moment(date).isBefore(moment(startDate))) {
      return true;
    }
    if (moment(endDate).isValid()
      && moment(date).isAfter(moment(endDate))) {
      return true;
    }
    return false;
  }

  render() {
    const {
      campaigns,
      startDate,
      endDate,
      focusedInput,
      pageWidth,
      toast,
    } = this.state;
    const showFullscreenCalendar = pageWidth < 768;
    return (
      <div className={styles.app}>
        <div className={styles.container}>
          <h1>Campaign Viewer</h1>
          <Panel>
            <PanelHeader className={styles.panelHeader}>
              <div className={styles.left}>
                <DateRangePicker
                  startDate={startDate}
                  startDateId="dateStart"
                  endDate={endDate}
                  endDateId="dateEnd"
                  focusedInput={focusedInput}
                  onDatesChange={this.onDatesChange}
                  onFocusChange={this.onFocusChange}
                  enableOutsideDays={false}
                  displayFormat="DD/MM/YYYY"
                  isOutsideRange={this.isOutsideRange}
                  numberOfMonths={showFullscreenCalendar ? 1 : 2}
                  withFullScreenPortal={showFullscreenCalendar}
                  showDefaultInputIcon={!showFullscreenCalendar}
                  block
                  readOnly
                  hideKeyboardShortcutsPanel
                  showClearDates
                  small
                />
              </div>
              <div className={styles.right}>
                <Input
                  placeholder="Search by Name"
                />
              </div>
            </PanelHeader>
            <PanelBody>
              <Table data={campaigns} />
            </PanelBody>
          </Panel>
        </div>
          <Toast
            {...toast}
          />
      </div>
    );
  }
}

export default App;
