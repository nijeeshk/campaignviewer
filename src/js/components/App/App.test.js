import React from 'react';
import moment from 'moment';
import { shallow, mount } from 'enzyme';
import App from './index';

jest.useFakeTimers();

const TEST_CAMPAIGNS = [
  {
    id: 1,
    name: 'campaign 1',
    startDate: '05/20/2019',
    endDate: '07/01/2019',
    Budget: 14587,
  },
  {
    id: 2,
    name: '2nd campaign',
    startDate: '01/20/2019',
    endDate: '02/01/2019',
    Budget: 14587,
  },
];

describe('App component', () => {
  it('renders without crashing', () => {
    shallow(<App />);
  });

  it('accepts campaigns and lists them correctly', () => {
    const app = shallow(<App />);
    app.instance().addCampaigns(TEST_CAMPAIGNS);
    expect(app).toMatchSnapshot();
  });

  it('filters campaigns when searched for', () => {
    const app = mount(<App />);
    app.instance().addCampaigns(TEST_CAMPAIGNS);
    app.find('input#searchInput').simulate('change', {
      target: { value: '2n' },
    });
    jest.advanceTimersByTime(1000);
    expect(app.state().filteredCampaigns).toEqual([
      {
        id: 2,
        name: '2nd campaign',
        startDate: '01/20/2019',
        endDate: '02/01/2019',
        Budget: 14587,
      },
    ]);
    app.unmount();
  });

  it('filters campaigns when date range updated', () => {
    const app = mount(<App />);
    app.instance().addCampaigns(TEST_CAMPAIGNS);
    app.instance().onDatesChange({
      startDate: moment('06/20/2019', 'MM/DD/YYYY'),
      endDate: moment('06/25/2019', 'MM/DD/YYYY'),
    });
    jest.advanceTimersByTime(1000);
    expect(app.state().filteredCampaigns).toEqual([
      {
        id: 1,
        name: 'campaign 1',
        startDate: '05/20/2019',
        endDate: '07/01/2019',
        Budget: 14587,
      },
    ]);
    app.unmount();
  });
});
