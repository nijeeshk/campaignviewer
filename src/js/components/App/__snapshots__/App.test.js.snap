// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component accepts campaigns and lists them correctly 1`] = `ShallowWrapper {}`;

exports[`App component filters campaigns when date range updated 1`] = `ReactWrapper {}`;

exports[`App component filters campaigns when searched for 1`] = `ReactWrapper {}`;
